Add parallel streams throughput tests, and enable them in the EGW workflow

giorio94 opened the pull request:
Extend the CLI performance tests to support measuring the throughput with multiple parallel streams, rather than a single one. This allows to bypass possible single flow limits due to resource exhaustion or cloud provider limits, such as \[1]. Unfortunately, netperf does not support multiple flows out of the box, hence we need to be a bit creative to run multiple instances and parse the resulting output. Enable the new test as part of the egress gateway scale test, and change the node instance type to m5n.xlarge, as it supports up to 25Gbps burst bandwidth.

Please review commit by commit, and refer to the individual commit messages for additional details.

\[1]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-network-bandwidth.html

giorio94 committed:
cli/perf: add throughput tests with multiple streams

Extend the CLI performance tests to support measuring the throughput
with multiple parallel streams, rather than a single one. This allows
to bypass possible single flow limits due to resource exhaustion or
cloud provider limits, such as [1]. Unfortunately, netperf does not
support multiple flows out of the box, hence we need to be a bit
creative to run multiple instances and parse the resulting output.

[1]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-network-bandwidth.html

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
gha/scale-egw: enable throughput tests with multiple streams

So that we circumvent the single flow limitations imposed by AWS [1].
Additionally, let's change the node instance type to m5n.xlarge, as
it supports up to 25Gbps burst bandwidth.

[1]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-network-bandwidth.html

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
gha/scale-egw: add sanity check to performance test

Gather the amount of traffic forwarded by the gateway node during the
network performance test, and assert that it is close to 0 in the
baseline test, and way higher in the EGW test. This serves as a sanity
check that traffic is indeed flowing through the egress gateway when
expected.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
gha/scale-egw: don't enable egress gateway in baseline test

Let's not enable the egress gateway functionality at all in the baseline
test, rather than enabling it but not configuring any policy. This
ensures that baseline Cilium CPU/memory usage are not influenced by any
Egress Gateway related tasks.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

christarazi commented on cilium-cli/connectivity/perf/benchmarks/netperf/perfpod.go:
```diff
@@ -204,10 +199,58 @@ func NetperfCmd(ctx context.Context, sip string, perfTest common.PerfTests, a *c
+	}
+	exec := buildExecCommand(test, sip, perfTest.Duration, args)
+
+	if streams >= 2 {
+		exec = []string{"/bin/bash", "-c",
+			// We write the output of each process to a separate file and cat them
+			// at the end to prevent the possibility of interleaved output.
+			fmt.Sprintf("DIR=$(mktemp -d); for i in {1..%d}; do %s > $DIR/out$i.out & done; wait; cat $DIR/*; rm -rf $DIR",
+				streams, strings.Join(exec, " "),
+			)}
+	}
```
Is it possible to do this in a more Go native way? We've tried to avoid bash "magic" in the CLI (although there isn't too much magic actually going on here, but rather it feels hacky) because it is error-prone.

giorio94 replied:
The main reason I've preferred doing this in bash rather than spawning multiple go routines each execing into the client pod is that the latter approach looks more fragile to me, in the sense of less likely to start all the parallel processes reasonably at the same time, due to the potential latency in establishing the exec stream, client/API server throttling, and so on. 

Possibly not mattering much on longer tests (say >10s) and low numbers of parallel streams, but more relevant on shorter ones and with higher numbers of parallel streams, due to the overall latency potentially being in the same order of magnitude. And definitely less predictable in general.

Another alternative could be hiding the logic in a script included in the docker image, but that seemed a bit overkill to me as it would require to regenerate the images. That would also make the logic here more obscure, given that it would call a custom command, rather than well known bash utils. Happy to take that path if you have strong opinions though.

marseel commented on cilium-cli/connectivity/perf/benchmarks/netperf/perfpod.go:
```diff
@@ -204,10 +199,58 @@ func NetperfCmd(ctx context.Context, sip string, perfTest common.PerfTests, a *c
+	lines := slices.DeleteFunc(
+		strings.Split(output, "\n"),
+		// Result lines always start with a number, hence drop all the others.
+		func(line string) bool { return len(line) == 0 || line[0] < '0' || line[0] > '9' },
```
I know I am guilty of it as well, but could we add a simple unit test for parsing netperf output :pleading_face: ?

giorio94 replied:
Added in https://github.com/cilium/cilium/pull/38027/commits/d6fbdca5a27537ea347eb3aa5369d10dc304aba5.

giorio94 committed:
cli/perf: add unit test covering the NetperfCmd function

Following the extension of this logic to additionally support multiple
parallel streams, let's add a unit test exercising the different paths,
for increased confidence in its correctness, and to prevent future
regressions.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

derailed reviewed (approved):
@giorio94 Nice! Do you have a sense of the before/after throughput gain
